### PR TITLE
Support `Sequence` as `List` in JIT

### LINKED ISSUE #103803: Support `Sequence` type in JIT
### 🚀 The feature, motivation and pitch

This is a sub-task of #103798.

We have functions annotated to accept `List[int]` but actually get `torch.Size` passed in, e.g.:
https://github.com/pytorch/pytorch/blob/918fe519a0f9bdb2dd213eaaffbcd420b7391280/torch/nn/functional.py#L2435
https://github.com/pytorch/pytorch/blob/918fe519a0f9bdb2dd213eaaffbcd420b7391280/torch/nn/functional.py#L2481

However, static type checkers like mypy does not like it because `torch.Size`, which is a subtype of `Tuple[int, ...]`, is not compatible with `List[int]`. This conflicts with JIT, which does not understand `Tuple[int, ...]` without a fixed length.

There are many cases of this problem, all involving functions taking in tensor sizes or similar (e.g., `Tensor.size` or `kernel_size`), where we only use them as `collections.abc.Sequence`.

Therefore, I propose to annotate them as `Sequence[int]` and ask JIT to interpret `Sequence[int]` as `List[int]`.

By annotating the sizes as `Sequence[int]`, we can i) make mypy happy because `List[int]` and `Tuple[int, ...]` (including `torch.Size`) are both `Sequence[int]`, and ii) borrow the JIT infrastructure for `ListType` which does not care about the length, bypassing the tuple length problem.

### Alternatives

1. _Adjust the implementation of `TupleType` in JIT to support `Tuple[int, ...]`_. This can be the best solution, but I don't know how to do it. Also, I assume it's not easy because otherwise it would already be implemented.
2. _Implement `SequenceType`, which differs from `ListType`_. This is also a good option, maybe second to the above, because it's safer than reusing the `ListType` -- some operations supported by `List` are not by `Sequence`, although JIT seems not to be using them.
3. _Add dozens of `type:ignore`s to suppress all related mypy warnings_. The good point is: the affected functions are for internal use (with `_` prefix) and will not affect public APIs (i.e., the users will not see any problem). However, this can leave the code error-prone.

### Additional context

Since we mark them as `List[int]` for now, even with `Tuple`s/`Size`s, directly reusing `List` for `Sequence` will not cause any problems. We actually keep how everything works under the hood, just switching to a different name that can satisfy mypy.

### PR DESCRIPTION
**Pull request by lkct**
Fixes #103803

The motivation is explained in the above issue. This is the easiest yet effective fix.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by lkct** [comment]
**To reviewers:**
About the test failure on public API, I'm copying the pattern of existing examples. I'm not sure if it's intended or if I missed something. Any help would be appreciated.

File: test/allowlist_for_publicAPI.json [hunk 1]
```diff
@@ -353,6 +353,7 @@
     "IntType",
     "List",
     "ListType",
+    "Sequence",
```
Comment (albanD) [inline test/allowlist_for_publicAPI.json:356]: You should not be adding anything to this file in general.

**Review by albanD — COMMENTED** [review]
I'm afraid no one is actively maintaining the jit frontend.
I'm not sure if anyone will be able to review this.

**Comment by lkct** [comment]
> I'm afraid no one is actively maintaining the jit frontend.
> I'm not sure if anyone will be able to review this.

That's not good news...
Are we going to deprecate JIT in favor of `torch.compile`?

File: test/allowlist_for_publicAPI.json (hunk 1 above)
Comment (lkct) [reply test/allowlist_for_publicAPI.json:356]: Actually, I'm not sure why to add this to public API. However, to follow the existing pattern I did this.

Should I just hide it from `__all__`?

**Review by davidberard98 — COMMENTED** [review]
Have you tested this change with code that uses Sequence annotations?

This change seems like it would be reasonable if this was the only change required... however, the torchscript type system is pretty complex and I would suspect that there would need to also be changes in the C++ parts to handle parsing the annotations.

Also - do you know whether torchscript can safely handle `Tuple[X, ...]` types as if they are `List[X]`? And is this a concern if we landed this change?

**Pull request closed without merging**